Take grant on WebKit extension process on launch

pvollan opened the pull request:
#### 094fb96f2ba5ca89933443b0f0922c6612cc5e50
<pre>
Take grant on WebKit extension process on launch
<a href="https://bugs.webkit.org/show_bug.cgi?id=269071">https://bugs.webkit.org/show_bug.cgi?id=269071</a>
<a href="https://rdar.apple.com/122647959">rdar://122647959</a>

Reviewed by Brent Fulgham and Chris Dumez.

There is currently a race condition on launch of WebKit extension processes where it is possible that
an extension process will be suspended before it is able to reply to the XPC bootstrap message. This is
because the system is only holding an initial assertion for a short period of time, and if the process
has not replied to the XPC bootstrap message within that timeframe, it can become suspended and will
then never reply to the bootstrap message, and the UI process will consider this process as hung.
This can be addressed by taking a grant on the process and releasing that grant when we have received
the XPC reply.

* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
(WebKit::ProcessLauncher::releaseLaunchGrant):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::LaunchGrant::LaunchGrant):
(WebKit::LaunchGrant::~LaunchGrant):
(WebKit::ProcessLauncher::launchProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didFinishLaunching):

Canonical link: <a href="https://commits.webkit.org/274390@main">https://commits.webkit.org/274390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f106326bbd7ba2ee15a333047f37604374d66142

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38957 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41310 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41491 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20764 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15238 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39530 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15069 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13086 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42768 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35369 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38881 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13769 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11370 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37104 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15375 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5083 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -190,7 +205,10 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
+
+        RefPtr<LaunchGrant> launchGrant = adoptRef(*new LaunchGrant(process));
+
+        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process), launchGrant = launchGrant] {
```
Should we use `WTFMove()` on the passed `launchGrant` here?

brentfulgham commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -208,6 +226,7 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
 
             launcher->m_xpcConnection = WTFMove(xpcConnection);
             launcher->m_process = WTFMove(process);
+            launcher->m_launchGrant = launchGrant;
```
Should we `WTFMove(launchGrant)` here?

pvollan replied:
Ah, yes, these are good point, I have addressed this in the latest patch.

Thanks for reviewing!

brentfulgham reviewed (approved):
Should we do anything in `ProcessLauncher::platformDestroy()` or `ProcessLauncher::platformInvalidate()` to clean up a pending LaunchGrant?

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -311,11 +311,16 @@ bool AuxiliaryProcessProxy::dispatchSyncMessage(IPC::Connection& connection, IPC
 
+#if USE(EXTENSIONKIT)
+    if (launcher)
+        launcher->releaseLaunchGrant();
```
We're releasing the grant too early here. The ProcessThrottler hasn't had a chance to take a process assertion yet, thus, the child process will likely suspend and resume shortly after.

The ProcessThrottler only gets a chance to take a process assertion after `ProcessThrottler::didConnectToProcess()` has been called (because that's when it gets the PID). `didConnectToProcess()` get called later on, in subclasses' didFinishLaunching().

We should probably refactor this code a bit so that we release the grant *after* the process throttler has had a chance to take an assertion. BTW, I think AuxiliaryProcessProxy could in theory notify the ProcessThrottler itself nowadays. It used to be that the AuxiliaryProcessProxy didn't know about the ProcessThrottler. However, I believe we changed that a while back.

pvollan replied:
Ah, good point! In order to make the changes minimal now, would you be ok with calling `releaseLaunchGrant` in each process proxy?

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -190,7 +205,10 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
         }
-        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process)] {
+
+        RefPtr<LaunchGrant> launchGrant = adoptRef(*new LaunchGrant(process));
```
`RefPtr<LaunchGrant>` -> `Ref`. Also, you shouldn't be doing this manually. Please add a static `create()` function to LaunchGrant and use that, like we always do (and for good reason). Also, please make the constructor private.

pvollan replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -190,7 +205,10 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
+
+        RefPtr<LaunchGrant> launchGrant = adoptRef(*new LaunchGrant(process));
+
+        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process), launchGrant = WTFMove(launchGrant)] {
```
`weakProcessLauncher = weakProcessLauncher` -> `weakProcessLauncher`
`name = name` -> `name`
`RetainPtr<_SEExtensionProcess>(process)` -> `retainPtr(process)`

cdumez commented on Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
```diff
@@ -57,6 +58,18 @@ enum class SandboxPermission {
+class LaunchGrant : public ThreadSafeRefCounted<LaunchGrant> {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    LaunchGrant(_SEExtensionProcess *);
```
Refcounted class should NEVER have public constructor, please add a static `create()` function and make the constructor private. It is too error-prone otherwise. Please mark constructor as explicit too.

pvollan replied:
Done!

cdumez commented on Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
```diff
@@ -57,6 +58,20 @@ enum class SandboxPermission {
 
+#if USE(EXTENSIONKIT)
+class LaunchGrant : public ThreadSafeRefCounted<LaunchGrant> {
+    WTF_MAKE_FAST_ALLOCATED;
```
Not necessary since you subclass ThreadSafeRefCounted

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -159,7 +178,8 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
     auto handler = [](ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, _SEExtensionProcess* process, ASCIILiteral name, NSError* error)
     {
-        if (!weakProcessLauncher.get()) {
+        auto launcher = weakProcessLauncher.get();
```
Please don't hide smart pointers behind `auto`. use `RefPtr` here.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -190,7 +210,10 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
         }
-        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process)] {
+
+        Ref<LaunchGrant> launchGrant = LaunchGrant::create(process);
```
`<LaunchGrant>` is unnecessary.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -190,7 +210,10 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
+
+        Ref<LaunchGrant> launchGrant = LaunchGrant::create(process);
+
+        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process), launchGrant = WTFMove(launchGrant)] () mutable {
```
Why are you still using ` foo = foo` here instead of just capturing `foo` ?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -190,7 +210,10 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
+        Ref<LaunchGrant> launchGrant = LaunchGrant::create(process);
+
+        callOnMainRunLoop([weakProcessLauncher = weakProcessLauncher, name = name, process = RetainPtr<_SEExtensionProcess>(process), launchGrant = WTFMove(launchGrant)] () mutable {
             auto launcher = weakProcessLauncher.get();
```
RefPtr

cdumez commented on Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
```diff
@@ -594,6 +594,10 @@ void GPUProcessProxy::didFinishLaunching(ProcessLauncher* launcher, IPC::Connect
     m_throttler.didConnectToProcess(*this);
+#if USE(EXTENSIONKIT)
+    if (launcher)
+        launcher->releaseLaunchGrant();
```
It is "okayish" to do this in subclasses for now. However, I think it'd be a lot less error-prone if we notified the process throttler and released the launch grant in the base class. I think we should consider refactoring in a follow-up at least.

This would have avoided you introducing a bug. You forgot to release the grant in `ModelProcessProxy::didFinishLaunching()`.

pvollan replied:
Ah, good catch, Chris! I also think refactoring this is the best path forward. I can look into that in a follow up. The motivation for not doing the refactoring now was to minimize the risk of regressions.

Thanks for reviewing!

cdumez replied:
Maybe we can add a FIXME comment at least.

cdumez reviewed (approved):
LGTM

pvollan committed:
Take grant on WebKit extension process on launch
https://bugs.webkit.org/show_bug.cgi?id=269071
rdar://122647959

Reviewed by Brent Fulgham and Chris Dumez.

There is currently a race condition on launch of WebKit extension processes where it is possible that
an extension process will be suspended before it is able to reply to the XPC bootstrap message. This is
because the system is only holding an initial assertion for a short period of time, and if the process
has not replied to the XPC bootstrap message within that timeframe, it can become suspended and will
then never reply to the bootstrap message, and the UI process will consider this process as hung.
This can be addressed by taking a grant on the process and releasing that grant when we have received
the XPC reply.

* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
(WebKit::ProcessLauncher::releaseLaunchGrant):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::LaunchGrant::LaunchGrant):
(WebKit::LaunchGrant::~LaunchGrant):
(WebKit::ProcessLauncher::launchProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didFinishLaunching):

Canonical link: https://commits.webkit.org/274390@main
